doc/compiler.md

```diff
@@ -0,0 +1,205 @@
+
+## Simple usage
+
+Let's take a simple ProtoDef definition and convert it to use the ProtoDef compiler:
+
+ProtoDef:
+```javascript
+const ProtoDef = require('protodef').ProtoDef
+
+// Create a ProtoDef instance
+const proto = new ProtoDef()
+proto.addTypes(require('./protocol.json'))
+
+// Encode and decode a message
+const buffer = proto.createPacketBuffer('mainType', result)
+const result = proto.parsePacketBuffer('mainType', buffer)
+```
+
+ProtoDef Compiler:
+```javascript
+const { ProtoDefCompiler } = require('protodef').Compiler
+
+// Create a ProtoDefCompiler instance
+const compiler = new ProtoDefCompiler()
+compiler.addTypesToCompile(require('./protocol.json'))
+
+// Compile a ProtoDef instance
+const compiledProto = await compiler.compileProtoDef()
+
+// Use it as if it were a normal ProtoDef
+const buffer = compiledProto.createPacketBuffer('mainType', result)
+const result = compiledProto.parsePacketBuffer('mainType', buffer)
+```
+
+## New datatypes
+
+Like the ProtoDef interpreter, the ProtoDef compiler can be extended with custom datatypes. To register a custom type, use the `addTypes(types)` method of the ProtoDef compiler. The `types` parameter is an object with the following structure:
+
+```javascript
+{
+  Read: {
+    'type1': ['native', /* implementation */],
+    'type2': ['context', /* implementation */],
+    'type3': ['parametrizable', /* implementation */],
+    /* ... */
+  },
+
+  Write: {
+    'type1': ['native', /* implementation */],
+    'type2': ['context', /* implementation */],
+    'type3': ['parametrizable', /* implementation */],
+    /* ... */
+  },
+
+  SizeOf: {
+    'type1': ['native', /* implementation */],
+    'type2': ['context', /* implementation */],
+    'type3': ['parametrizable', /* implementation */],
+    /* ... */
+  }
+}
+```
+
+The types can be divided into 3 categories:
+
+### Native Type
+
+A native type is a type read or written by a function that will be called in its original context. Use this when you need access to external definitions.
+
+Example:
+```javascript
+const UUID = require('uuid-1345')
+
+{
+  Read: {
+    'UUID': ['native', (buffer, offset) => {
+      return {
+        value: UUID.stringify(buffer.slice(offset, 16 + offset)), // A native type can access all captured definitions
+        size: 16
+      }
+    }]
+  },
+  Write: {
+    'UUID': ['native', (value, buffer, offset) => {
+      const buf = UUID.parse(value)
+      buf.copy(buffer, offset)
+      return offset + 16
+    }]
+  },
+  SizeOf: {
+    'UUID': ['native', 16] // For SizeOf, a native type can be a function or directly an integer
+  }
+}
+```
+
+### Context Type
+
+A context type is a type that will be called in the protocol's context. It can refer to registred native types using `native.{type}()` or context types (provided and generated) using `ctx.{type}()`, but cannot access its original context.
```

Comment:
this is similar to what happens in the interpreter ? (might be useful to mention it)